### PR TITLE
Add Beta-Binomial distribution for modeling scBS-seq data

### PR DESCRIPTION
**Pull request by ethanweinberger**
New assays (e.g. snmC-seq) have emerged for measuring DNA methylation using bisulfite sequencing (BS-seq). Due to issues with coverage (i.e., different cytosines are measured for each cell), BS-seq measurements at individual cytosines are often aggregated across larger region features (e.g. gene bodies, enhancers, etc.) before performing computational analyses. Due to observed overdispersion, when modeling these sums of binary cytosine-level measurements, the Beta-Binomial distribution is commonly used ([1](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02329-8), [2](https://openreview.net/forum?id=Mg2DM0F3AY)).

To facilitate prototyping future models of BS-seq data, this PR adds the Beta-Binomial with mean-dispersion parameterization to the scvi-tools collection of distributions.

@martinkim0 @canergen As discussed previously I added the Beta-Binomial distribution to the main `scvi` folder (as opposed to external). Let me know what you'd like to see in terms of testing: because the class is just a thin wrapper around Pyro's Beta-Binomial distribution, I wasn't sure what would be appropriate here.

### CODE REVIEW

File: scvi/distributions/_constraints.py
Comment (canergen) [inline scvi/distributions/_constraints.py:1]: Can we use here the pytorch distribution constraints.interval?

**Comment by canergen** [comment]
LGTM. Can you add tests for the function using both parametrizations? I left two comments.

File: scvi/distributions/_constraints.py
Comment (ethanweinberger) [reply scvi/distributions/_constraints.py:1]: For the $\mu, \gamma$ Beta-Binomial parameterization both parameters lie within the open interval $(0, 1)$ (see https://rpubs.com/cakapourani/beta-binomial for reference). Unfortunately pytorch only provides constraints for closed intervals and half-open intervals, but not fully open intervals.

**Comment by ethanweinberger** [comment]
> LGTM. Can you add tests for the function using both parametrizations? I left two comments.

Happy to add tests, but what tests are you thinking specifically? Something like: likelihood values are the same for an $(\alpha, \beta)$ pair and the corresponding $(\mu, \gamma)$ pair?

**Comment by canergen** [comment]
I would check that log_probability of both parametrizations is close. It'd be also sufficient though to test the ValueError (set all parameters to None) and param1 (just check that you can sample from the distribution and it has a finite log_probability). I assume both parts of the code are not reached in methylVI.

**Comment by ethanweinberger** [comment]
> I would check that log_probability of both parametrizations is close. It'd be also sufficient though to test the ValueError (set all parameters to None) and param1 (just check that you can sample from the distribution and it has a finite log_probability). I assume both parts of the code are not reached in methylVI.

Added tests checking:

1. log_prob of both parameterizations is close, and
2. ValueError is raised if a parameterization is not specified

File: scvi/distributions/_beta_binomial.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            alpha, beta = broadcast_all(alpha, beta)
+        else:
+            mu, gamma = broadcast_all(mu, gamma)
+
+            alpha = mu * (1 - gamma) / gamma
+            beta = (mu - 1) * (gamma - 1) / gamma
+
+            # Clamping to force alpha, beta > 0 due to previously observed
+            # numerical stability issues.
+            alpha = torch.clamp(alpha, min=self._eps)
```
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply scvi/distributions/_beta_binomial.py:87]: I tried out your suggestion here, but the distortion introduced by softplus can lead to pretty large differences in log_prob between the $(\alpha, \beta)$ vs an equivalent $(\mu, \gamma)$ paramterization. I thus chose to leave this as-is unless you have a strong opinion here.

File: scvi/distributions/_beta_binomial.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import torch
+from pyro.distributions import BetaBinomial as BetaBinomialDistribution
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all
+
+from ._constraints import open_interval, optional_constraint
+
+
+class BetaBinomial(BetaBinomialDistribution):
+    r"""Beta binomial distribution.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:12]:
```suggestion
    r"""``BETA`` Beta binomial distribution.
```
Let's tag this as beta for now.

File: scvi/distributions/_beta_binomial.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    One of the following parameterizations must be provided:
+
+    (1), (`alpha`, `beta`, `total_counts`) where alpha and beta are the shape parameters of
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:16]:
```suggestion
    1. (`alpha`, `beta`, `total_counts`) where alpha and beta are the shape parameters of 
```

File: scvi/distributions/_beta_binomial.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    the beta distribution and total_counts is the number of trials.
+    (2), (`mu`, `gamma`, `total_counts`), which is the one used by methylVI. These
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:18]:
```suggestion

    2. (`mu`, `gamma`, `total_counts`), which is the one used by methylVI. These
```

File: scvi/distributions/_constraints.py [hunk 5]
```diff
@@ -18,3 +18,24 @@ def __repr__(self) -> str:
 def optional_constraint(constraint: Constraint) -> Constraint:
     """Returns a wrapped constraint that allows optional values."""
     return _Optional(constraint)
+
+
+class _OpenInterval(Constraint):
+    """Constrain to a real interval `(lower_bound, upper_bound)`"""
+
+    def __init__(self, lower_bound, upper_bound):
```
Comment (martinkim0) [inline scvi/distributions/_constraints.py:26]: Needs typing

File: scvi/distributions/_constraints.py [hunk 6]
```diff
@@ -18,3 +18,24 @@ def __repr__(self) -> str:
...
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        super().__init__()
+
+    def check(self, value):
```
Comment (martinkim0) [inline scvi/distributions/_constraints.py:31]: Needs typing

File: tests/distributions/test_constraints.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    # Within the interval is ok
+    assert constraint.check(0.5)
+
+    # Values outside interval are not ok
+    assert not constraint.check(1.1)
+    assert not constraint.check(-0.1)
+
+    # Endpoints are also not ok
+    assert not constraint.check(0)
+    assert not constraint.check(1.0)
```
Comment (martinkim0) [inline tests/distributions/test_constraints.py:19]: nit: call __repr__ here (just for test coverage)

File: scvi/distributions/_beta_binomial.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinkim0) [reply scvi/distributions/_beta_binomial.py:87]: This seems fine to me but should ideally be documented, maybe under a `Notes` section in the class docstring.

File: scvi/distributions/_beta_binomial.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    beta
+        Second shape parameter of the beta distribution.
+    mu
+        Mean of the distribution.
+    gamma
+        Dispersion.
+    validate_args
+        Raise ValueError if arguments do not match constraints
+    eps
+        Numerical stability constant
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:43]: Probably good to specifically reference how this is used, e.g. "Numerical stability constant used for minimum clamping when the second parameterization is specified."

File: scvi/distributions/_beta_binomial.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    3. :math:`y = \sum_{i}y_i`
+
+    Parameters
+    ----------
+    total_count
+        Number of trials.
+    alpha
+        First shape parameter of the beta distribution.
+    beta
+        Second shape parameter of the beta distribution.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:35]: Ideally we reference the parent Pyro class here and mention that alpha and beta refer to that parameterization.

File: scvi/distributions/_beta_binomial.py (hunk 8 above)
Comment (martinkim0) [reply scvi/distributions/_beta_binomial.py:43]: You can also put this under a `Notes` section in the docstring.

File: scvi/distributions/_beta_binomial.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+        self.gamma = gamma
+        self.alpha = alpha
+        self.beta = beta
+
+        super().__init__(
+            concentration1=alpha,
+            concentration0=beta,
+            total_count=total_count,
+            validate_args=validate_args,
+        )
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:100]: Need to override the parent `__repr__` since it uses optional parameters (see #2395)

File: tests/distributions/test_beta_binomial.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+from scvi.distributions import BetaBinomial
+
+
+def test_beta_binomial_distribution():
+    alpha = torch.ones(size=(2,))
+    beta = torch.ones_like(alpha)
+
+    num_trials = 100
+    total_count = torch.ones_like(alpha) * num_trials
+    x = torch.randint_like(total_count, high=num_trials)
```
Comment (martinkim0) [inline tests/distributions/test_beta_binomial.py:13]: Would be great to also see a call to `log_prob` with an `x` outside of the distribution's support.

File: tests/distributions/test_beta_binomial.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    mu = alpha / (alpha + beta)
+    gamma = 1 / (alpha + beta + 1)
+
+    dist_param_2 = BetaBinomial(total_count=total_count, mu=mu, gamma=gamma)
+
+    log_p_mu_gamma = dist_param_2.log_prob(x)
+    assert (log_p_alpha_beta - log_p_mu_gamma).abs().max().item() <= 1e-8
+
+    # Should fail as no parameterization is specified
+    with pytest.raises(ValueError):
```
Comment (martinkim0) [inline tests/distributions/test_beta_binomial.py:28]: also need to test for `ValueError` when both parameterizations are provided as well as overlaps (e.g. `total_counts`, `beta`, and `gamma` are provided).

**Review by martinkim0 — COMMENTED** [review]
Thanks @ethanweinberger, overall this looks great. I left some minor comments and nitpicks regarding documentation, small code changes, and tests.

**Comment by martinkim0** [comment]
Forgot to mention in my review: it would also be nice to have a release note under `CHANGELOG.md`. Thanks!

File: tests/distributions/test_beta_binomial.py (hunk 12 above)
Comment (ethanweinberger) [reply tests/distributions/test_beta_binomial.py:34]: Added a corresponding test

File: tests/distributions/test_beta_binomial.py (hunk 11 above)
Comment (ethanweinberger) [reply tests/distributions/test_beta_binomial.py:13]: Added a corresponding test

File: scvi/distributions/_beta_binomial.py (hunk 9 above)
Comment (ethanweinberger) [reply scvi/distributions/_beta_binomial.py]: Modified the docstring accordingly

File: scvi/distributions/_beta_binomial.py (hunk 8 above)
Comment (ethanweinberger) [reply scvi/distributions/_beta_binomial.py]: Added additional details as requested

File: scvi/distributions/_constraints.py (hunk 6 above)
Comment (ethanweinberger) [reply scvi/distributions/_constraints.py]: Done

File: scvi/distributions/_beta_binomial.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply scvi/distributions/_beta_binomial.py:105]: Added a corresponding `Notes` section as requested

**Comment by ethanweinberger** [comment]
@martinkim0 just finished addressing your comments. Let me know if there's anything else you'd like from me on this PR

File: scvi/distributions/_beta_binomial.py (hunk 10 above)
Comment (martinkim0) [reply scvi/distributions/_beta_binomial.py:118]: Yes, good idea. I had a PR open in the pytorch repo for this, I'll probably revisit it and then we can remove it for our distributions.

File: scvi/distributions/_beta_binomial.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+from ._constraints import open_interval, optional_constraint
+
+
+class BetaBinomial(BetaBinomialDistribution):
+    r"""``BETA`` Beta binomial distribution.
+
+    One of the following parameterizations must be provided:
+
+    1. (`alpha`, `beta`, `total_counts`) where alpha and beta are the shape parameters of
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:16]:
```suggestion
    1. (``alpha``, ``beta``, ``total_counts``) where alpha and beta are the shape parameters of
```

File: scvi/distributions/_beta_binomial.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+class BetaBinomial(BetaBinomialDistribution):
+    r"""``BETA`` Beta binomial distribution.
+
+    One of the following parameterizations must be provided:
+
+    1. (``alpha``, ``beta``, ``total_counts``) where alpha and beta are the shape parameters of
+    the beta distribution and total_counts is the number of trials.
+
+    2. (`mu`, `gamma`, `total_counts`), which is the one used by methylVI. These
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:19]:
```suggestion
    2. (``mu``, ``gamma``, ``total_counts``), which is the one used by methylVI. These
```

File: scvi/distributions/_beta_binomial.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+    One of the following parameterizations must be provided:
+
+    1. (``alpha``, ``beta``, ``total_counts``) where alpha and beta are the shape parameters of
+    the beta distribution and total_counts is the number of trials.
+
+    2. (``mu``, ``gamma``, ``total_counts``), which is the one used by methylVI. These
+    parameters respectively control the mean and dispersion of the distribution.
+
+    In the (`mu`, `gamma`) parameterization, samples from the beta-binomial are generated
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:22]:
```suggestion
    In the (``mu``, ``gamma``) parameterization, samples from the beta-binomial are generated
```

File: scvi/distributions/_beta_binomial.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+    gamma
+        Dispersion.
+    validate_args
+        Raise ValueError if arguments do not match constraints
+    eps
+        Numerical stability constant. (See Notes)
+
+    Notes
+    -----
+    Under the hood we use Pyro's `BetaBinomialDistribution` class to implement
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:52]:
```suggestion
    Under the hood we use :class:`~pyro.distributions.BetaBinomial` to implement
```

File: scvi/distributions/_beta_binomial.py [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    1. :math:`p_i \sim \textrm{Beta}(\mu, \gamma)`
+    2. :math:`y_i \sim \textrm{Ber}(p_i)`
+    3. :math:`y = \sum_{i}y_i`
+
+    Parameters
+    ----------
+    total_count
+        Number of trials.
+    alpha
+        As in the Pyro `BetaBinomial` distribution class,
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:34]:
```suggestion
        As in :class:`~pyro.distributions.BetaBinomial`,
```

File: scvi/distributions/_beta_binomial.py [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+    Parameters
+    ----------
+    total_count
+        Number of trials.
+    alpha
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the first shape parameterization of the beta
+        distribution.
+    beta
+        As in the Pyro `BetaBinomial` distribution class,
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:38]:
```suggestion
        As in :class:`~pyro.distributions.BetaBinomial`,
```

File: scvi/distributions/_beta_binomial.py [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+    as follows:
+
+    1. :math:`p_i \sim \textrm{Beta}(\mu, \gamma)`
+    2. :math:`y_i \sim \textrm{Ber}(p_i)`
+    3. :math:`y = \sum_{i}y_i`
+
+    Parameters
+    ----------
+    total_count
+        Number of trials.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:32]:
```suggestion
        Number of trials. Must be a non-negative integer.
```

File: scvi/distributions/_beta_binomial.py [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+    3. :math:`y = \sum_{i}y_i`
+
+    Parameters
+    ----------
+    total_count
+        Number of trials. Must be a non-negative integer.
+    alpha
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the first shape parameterization of the beta
+        distribution.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:36]:
```suggestion
        distribution. Must be greater than ``0``.
```

File: scvi/distributions/_beta_binomial.py [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+    total_count
+        Number of trials. Must be a non-negative integer.
+    alpha
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the first shape parameterization of the beta
+        distribution. Must be greater than ``0``.
+    beta
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the second shape parameterization of the beta
+        distribution.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:40]:
```suggestion
        distribution. Must be greater than ``0``.
```

File: scvi/distributions/_beta_binomial.py [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+    alpha
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the first shape parameterization of the beta
+        distribution. Must be greater than ``0``.
+    beta
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the second shape parameterization of the beta
+        distribution. Must be greater than ``0``.
+    mu
+        Mean of the distribution.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:42]:
```suggestion
        Mean of the distribution. Must be within the interval ``(0, 1)``.
```

File: scvi/distributions/_beta_binomial.py [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+        serves as the first shape parameterization of the beta
+        distribution. Must be greater than ``0``.
+    beta
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the second shape parameterization of the beta
+        distribution. Must be greater than ``0``.
+    mu
+        Mean of the distribution. Must be within the interval ``(0, 1)``.
+    gamma
+        Dispersion.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:44]:
```suggestion
        Dispersion. Must be within the interval ``(0, 1)``.
```

File: scvi/distributions/_beta_binomial.py [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+    beta
+        As in :class:`~pyro.distributions.BetaBinomial`,
+        serves as the second shape parameterization of the beta
+        distribution. Must be greater than ``0``.
+    mu
+        Mean of the distribution. Must be within the interval ``(0, 1)``.
+    gamma
+        Dispersion. Must be within the interval ``(0, 1)``.
+    validate_args
+        Raise ValueError if arguments do not match constraints
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:46]:
```suggestion
        Raise ``ValueError`` if arguments do not match the constraints.
```

File: scvi/distributions/_beta_binomial.py [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+        Dispersion. Must be within the interval ``(0, 1)``.
+    validate_args
+        Raise ``ValueError`` if arguments do not match the constraints.
+    eps
+        Numerical stability constant. (See Notes)
+
+    Notes
+    -----
+    Under the hood we use :class:`~pyro.distributions.BetaBinomial` to implement
+    the Beta-Binomial distribution. Thus, when the user specifies a (`mu`, `gamma`)
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:53]:
```suggestion
    the Beta-Binomial distribution. Thus, when the user specifies a (``mu``, ``gamma``)
```

File: scvi/distributions/_beta_binomial.py [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+    validate_args
+        Raise ``ValueError`` if arguments do not match the constraints.
+    eps
+        Numerical stability constant. (See Notes)
+
+    Notes
+    -----
+    Under the hood we use :class:`~pyro.distributions.BetaBinomial` to implement
+    the Beta-Binomial distribution. Thus, when the user specifies a (``mu``, ``gamma``)
+    parameterization, we must convert to the (`alpha`, `beta`) parameterization
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:54]:
```suggestion
    parameterization, we must convert to the (``alpha``, ``beta``) parameterization
```

File: scvi/distributions/_beta_binomial.py [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+    eps
+        Numerical stability constant. (See Notes)
+
+    Notes
+    -----
+    Under the hood we use :class:`~pyro.distributions.BetaBinomial` to implement
+    the Beta-Binomial distribution. Thus, when the user specifies a (``mu``, ``gamma``)
+    parameterization, we must convert to the (``alpha``, ``beta``) parameterization
+    used by the underlying Pyro distribution class. During this process, numerical
+    stability issues sometimes cause `alpha` or `beta` to be equal to (exactly) zero.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:56]:
```suggestion
    stability issues sometimes cause ``alpha`` or ``beta`` to be equal to (exactly) zero.
```

File: scvi/distributions/_beta_binomial.py [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+        Numerical stability constant. (See Notes)
+
+    Notes
+    -----
+    Under the hood we use :class:`~pyro.distributions.BetaBinomial` to implement
+    the Beta-Binomial distribution. Thus, when the user specifies a (``mu``, ``gamma``)
+    parameterization, we must convert to the (``alpha``, ``beta``) parameterization
+    used by the underlying Pyro distribution class. During this process, numerical
+    stability issues sometimes cause ``alpha`` or ``beta`` to be equal to (exactly) zero.
+    This is not allowed (`alpha` and `beta` must be strictly >0), so we clamp these
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:57]:
```suggestion
    This is not allowed (``alpha`` and ``beta`` must be strictly greater than 0), so we clamp these
```

File: scvi/distributions/_beta_binomial.py [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+
+    Notes
+    -----
+    Under the hood we use :class:`~pyro.distributions.BetaBinomial` to implement
+    the Beta-Binomial distribution. Thus, when the user specifies a (``mu``, ``gamma``)
+    parameterization, we must convert to the (``alpha``, ``beta``) parameterization
+    used by the underlying Pyro distribution class. During this process, numerical
+    stability issues sometimes cause ``alpha`` or ``beta`` to be equal to (exactly) zero.
+    This is not allowed (``alpha`` and ``beta`` must be strictly greater than 0), so we clamp these
+    values to be greater than a small constant `eps`.
```
Comment (martinkim0) [inline scvi/distributions/_beta_binomial.py:58]:
```suggestion
    values to be greater than a small constant ``eps``.
```

File: scvi/distributions/_constraints.py [hunk 30]
```diff
@@ -18,3 +18,24 @@ def __repr__(self) -> str:
 def optional_constraint(constraint: Constraint) -> Constraint:
     """Returns a wrapped constraint that allows optional values."""
     return _Optional(constraint)
+
+
+class _OpenInterval(Constraint):
+    """Constrain to a real interval `(lower_bound, upper_bound)`"""
```
Comment (martinkim0) [inline scvi/distributions/_constraints.py:24]:
```suggestion
    """Constrain to a real interval ``(lower_bound, upper_bound)``"""
```

**Review by martinkim0 — APPROVED** [review]
Thanks for addressing the comments! I added a couple of changes for formatting of the docstrings. 

I think this is ready to merge once tests + docs pass.
